### PR TITLE
feat(http): add basic http service

### LINKED ISSUE #2028: land basic http lib
Scope: Should perform basic request/response, returning `EventEmitter` object, using XHR, with mock backend.

Example:

```js
http('https://google.com').subscribe((res) => console.log(res));
```

### PR DESCRIPTION
**Pull request by jeffbcross**
TODO:
- [x] Add documentation comments
- [x] Open issues to capture remaining features for current milestone

This implementation only works in JavaScript, while the Observable transpilation
story gets worked out. Right now, the service just makes a simple request,
and returns an Observable of Response.

Fixes #2028

### CODE REVIEW

File: modules/angular2/http.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import {bind, Binding} from 'angular2/di';
+import {Http, HttpFactory} from './src/http/http';
+import {XHRBackend} from 'angular2/src/http/backends/xhr_backend';
+import {BrowserXHR} from 'angular2/src/http/backends/browser_xhr';
+
+export var httpBindings: List<any> = [
```
Comment (gdi2290) [inline modules/angular2/http.ts]: the pattern seems to be Injectables
https://github.com/angular/angular/blob/master/modules/angular2/router.ts#L33
https://github.com/angular/angular/blob/master/modules/angular2/router.ts#L33

File: modules/angular2/src/http/base_request_options.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    this.headers = headers;
+    this.body = body;
+    this.mode = mode;
+    this.credentials = credentials;
+    this.cache = cache;
+  }
+}
+;
+
+export var baseRequestOptions = Object.freeze(new BaseRequestOptions());
```
Comment (gdi2290) [inline modules/angular2/src/http/base_request_options.ts]: Should this be `const` and `CONST_EXPR` also adding `@CONST` to `BaseRequestOptions` if you want to freeze it (for Dart etc)

File: modules/examples/src/http/rx_pipe.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
+/// <reference path="../../../angular2/typings/rx/rx.all.d.ts" />
+
+import {isBlank, isPresent, CONST} from 'angular2/src/facade/lang';
+import {Observable, ObservableWrapper} from 'angular2/src/facade/async';
+import {Pipe, WrappedValue, PipeFactory} from 'angular2/src/change_detection/pipes/pipe';
+import {ChangeDetectorRef} from 'angular2/src/change_detection/change_detector_ref';
+import * as Rx from 'rx';
+
+export class RxPipe extends Pipe {
```
Comment (gdi2290) [inline modules/examples/src/http/rx_pipe.ts]: this can be simpler
https://gist.github.com/gdi2290/e9b2880a1d13057197d7

**Comment by jeffbcross** [comment]
Thanks for all the comments, @gdi2290!

File: modules/angular2/src/http/backends/browser_xhr.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+declare var window;
+
+import {Injectable} from 'angular2/di';
+
+// Make sure not to evaluate this in a non-browser environment!
+@Injectable()
+export class BrowserXHR {
+  constructor(args: any) { return <any>(new window.XMLHttpRequest(args)); }
```
Comment (jeffbcross) [inline modules/angular2/src/http/backends/browser_xhr.ts]: remove args for symmetry

File: modules/angular2/src/http/backends/xhr_backend.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import {Injectable} from 'angular2/di';
+import {BrowserXHR} from './browser_xhr';
+import * as Rx from 'rx';
+
+export class XHRConnection implements IConnection {
+  request: Request;
+  response: Rx.Subject<Response>;
+  readyState: ReadyStates;
+  private _xhr;
+  constructor(req: Request, nativeConstruct: any) {
```
Comment (jeffbcross) [inline modules/angular2/src/http/backends/xhr_backend.ts]: Why <any>? And nativeConstruct should have different casing.

File: modules/angular2/src/http/backends/xhr_backend.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    this._xhr.send(this.request.body);
+  }
+
+  dispose(): void { this._xhr.abort(); }
+}
+
+@Injectable()
+export class XHRBackend implements IConnectionBackend {
+  private _nativeConstruct: BrowserXHR;
+  constructor(@Inject(BrowserXHR) nativeConstruct: BrowserXHR) {
```
Comment (jeffbcross) [inline modules/angular2/src/http/backends/xhr_backend.ts]: Use either `@Inject` or type, not both.

File: modules/angular2/src/http/backends/xhr_backend.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        () => {this.response.onNext(new Response(this._xhr.response || this._xhr.responseText))});
+    this._xhr.send(this.request.body);
+  }
+
+  dispose(): void { this._xhr.abort(); }
+}
+
+@Injectable()
+export class XHRBackend implements IConnectionBackend {
+  private _nativeConstruct: BrowserXHR;
```
Comment (jeffbcross) [inline modules/angular2/src/http/backends/xhr_backend.ts]: redundant to constructor args

File: modules/angular2/src/http/backends/xhr_backend.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  constructor(@Inject(BrowserXHR) nativeConstruct: BrowserXHR) {
+    this._nativeConstruct = nativeConstruct;
+  }
+  createConnection(request: Request): IConnection {
```
Comment (jeffbcross) [inline modules/angular2/src/http/backends/xhr_backend.ts]: Re-consider I-prefix of interface, be consistent with rest of framework.

File: modules/angular2/src/http/static_response.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  blob(): Blob {
+    throw new BaseException('"blob()" method not implemented on Response superclass');
+  }
+
+  json(): JSON {
+    var parsed;
+    if (isJsObject(this.body)) {
+      return <JSON>this.body;
+    } else if (isString(this.body)) {
+      try {
```
Comment (jeffbcross) [inline modules/angular2/src/http/static_response.ts]: Get rid of try/catch, default to fetch spec behavior

File: modules/angular2/src/http/static_request.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import {RequestMethods, RequestModesOpts, RequestCredentialsOpts} from './enums';
+import {URLSearchParams} from './url_search_params';
+import {IRequestOptions, IRequest} from './interfaces';
+import {Headers} from './headers';
+import {baseRequestOptions} from './base_request_options';
+import {BaseException, RegExpWrapper} from 'angular2/src/facade/lang';
+
+export class Request implements IRequest {
+  method: RequestMethods;
+  mode: RequestModesOpts;
```
Comment (jeffbcross) [inline modules/angular2/src/http/static_request.ts]: TODO: determine how this will be implemented

File: modules/angular2/src/http/http.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+var Observable;
+if (Rx.hasOwnProperty('default')) {
+  Observable = (<any>Rx).default.Rx.Observable;
+} else {
+  Observable = Rx.Observable;
+}
+export function HttpFactory(backend: XHRBackend) {
+  return function(url: string, options?: IRequestOptions) {
```
Comment (jeffbcross) [inline modules/angular2/src/http/http.ts]: Move instantiation of base_response_options here, and make it injectable.

**Comment by jeffbcross** [comment]
@vsavkin did you have any other feedback other than what we captured on our call? (If not, can you add lgtm label?)

**Comment by vsavkin** [comment]
@jeffbcross no, just make Travis happy!

File: modules/angular2/src/http/static_request.ts (hunk 10 above)
Comment (robwormald) [inline modules/angular2/src/http/static_request.ts]: realllly picky but you're switching between "Opts" and "Options" here?

**Comment by jeffbcross** [comment]
Merged at 21568106b114943b59ce37832d82c8fb9a63285b

**Pull request closed without merging**